### PR TITLE
r/iam_instance_profile - add tagging support to instance profile + sweeper

### PR DESCRIPTION
**Pull request by DrFaust92**
### Community Note

* Please vote on this pull request by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original pull request comment to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for pull request followers and do not help prioritize the request

Relates #17594

Output from acceptance testing:

```
$ make testacc TESTARGS='-run=TestAccAWSIAMInstanceProfile_'
--- PASS: TestAccAWSIAMInstanceProfile_withoutRole (32.05s)
--- PASS: TestAccAWSIAMInstanceProfile_disappears_role (36.93s)
--- PASS: TestAccAWSIAMInstanceProfile_disappears (37.08s)
--- PASS: TestAccAWSIAMInstanceProfile_basic (42.21s)
--- PASS: TestAccAWSIAMInstanceProfile_namePrefix (48.56s)
--- PASS: TestAccAWSIAMInstanceProfile_tags (98.49s)
```

### CODE REVIEW

**Comment by ewbankkit** [comment]
Verified acceptance tests:

```console
$ make testacc TEST=./aws/ TESTARGS='-run=TestAccAWSIAMInstanceProfile_'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws -v -count 1 -parallel 20 -run=TestAccAWSIAMInstanceProfile_ -timeout 120m
=== RUN   TestAccAWSIAMInstanceProfile_basic
=== PAUSE TestAccAWSIAMInstanceProfile_basic
=== RUN   TestAccAWSIAMInstanceProfile_withoutRole
=== PAUSE TestAccAWSIAMInstanceProfile_withoutRole
=== RUN   TestAccAWSIAMInstanceProfile_tags
=== PAUSE TestAccAWSIAMInstanceProfile_tags
=== RUN   TestAccAWSIAMInstanceProfile_namePrefix
=== PAUSE TestAccAWSIAMInstanceProfile_namePrefix
=== RUN   TestAccAWSIAMInstanceProfile_disappears
=== PAUSE TestAccAWSIAMInstanceProfile_disappears
=== RUN   TestAccAWSIAMInstanceProfile_disappears_role
=== PAUSE TestAccAWSIAMInstanceProfile_disappears_role
=== CONT  TestAccAWSIAMInstanceProfile_basic
=== CONT  TestAccAWSIAMInstanceProfile_disappears
=== CONT  TestAccAWSIAMInstanceProfile_disappears_role
=== CONT  TestAccAWSIAMInstanceProfile_tags
=== CONT  TestAccAWSIAMInstanceProfile_namePrefix
=== CONT  TestAccAWSIAMInstanceProfile_withoutRole
--- PASS: TestAccAWSIAMInstanceProfile_disappears (22.11s)
--- PASS: TestAccAWSIAMInstanceProfile_withoutRole (22.58s)
--- PASS: TestAccAWSIAMInstanceProfile_disappears_role (22.69s)
--- PASS: TestAccAWSIAMInstanceProfile_basic (25.18s)
--- PASS: TestAccAWSIAMInstanceProfile_namePrefix (26.87s)
--- PASS: TestAccAWSIAMInstanceProfile_tags (41.52s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	41.641s
```

**Comment by ewbankkit** [comment]
Running the sweeper:

```console
$ TEST=./aws SWEEP=us-west-2 SWEEPARGS=-sweep-run=aws_iam_instance_profile make sweep
WARNING: This will destroy infrastructure. Use only in development accounts.
go test ./aws -v -sweep=us-west-2 -sweep-run=aws_iam_instance_profile -timeout 60m
2021/03/06 16:39:38 [DEBUG] Running Sweepers for region (us-west-2):
2021/03/06 16:39:38 [DEBUG] Running Sweeper (aws_iam_instance_profile) in region (us-west-2)
2021/03/06 16:39:38 [INFO] AWS Auth provider used: "EnvProvider"
2021/03/06 16:39:38 [DEBUG] Trying to get account information via sts:GetCallerIdentity
2021/03/06 16:39:38 [DEBUG] Trying to get account information via sts:GetCallerIdentity
2021/03/06 16:39:39 [ERROR] error deleting IAM Instance Profile (kit-assumable-role): deleting IAM instance profile kit-assumable-role: DeleteConflict: Cannot delete entity, must remove roles from instance profile first.
	status code: 409, request id: ef65b42d-54c2-4085-8e2f-82ebec2f9d12
2021/03/06 16:39:39 [ERROR] Error running Sweeper (aws_iam_instance_profile) in region (us-west-2): 1 error occurred:
	* error deleting IAM Instance Profile (kit-assumable-role): deleting IAM instance profile kit-assumable-role: DeleteConflict: Cannot delete entity, must remove roles from instance profile first.
	status code: 409, request id: ef65b42d-54c2-4085-8e2f-82ebec2f9d12

FAIL	github.com/terraform-providers/terraform-provider-aws/aws	1.241s
FAIL
GNUmakefile:20: recipe for target 'sweep' failed
make: *** [sweep] Error 1
```

I think you'll need to add `Dependencies: []string{"aws_iam_role"}.` to the `resource.Sweeper`.

**Comment by DrFaust92** [comment]
Thanks! added sweeper dep.

File: aws/resource_aws_iam_instance_profile_test.go [hunk 1]
```diff
@@ -67,6 +117,51 @@ func TestAccAWSIAMInstanceProfile_withoutRole(t *testing.T) {
...
 
+func TestAccAWSIAMInstanceProfile_tags(t *testing.T) {
+	var conf iam.GetInstanceProfileOutput
+	resourceName := "aws_iam_instance_profile.test"
+	rName := acctest.RandString(5)
+
+	resource.ParallelTest(t, resource.TestCase{
+		PreCheck:     func() { testAccPreCheck(t) },
+		Providers:    testAccProviders,
+		CheckDestroy: testAccCheckAWSInstanceProfileDestroy,
```
Comment (bflad) [inline aws/resource_aws_iam_instance_profile_test.go:128]:
```suggestion
		PreCheck:     func() { testAccPreCheck(t) },
		ErrorCheck:   testAccErrorCheck(t, iam.EndpointsID),
		Providers:    testAccProviders,
		CheckDestroy: testAccCheckAWSInstanceProfileDestroy,
```

File: website/docs/r/iam_instance_profile.html.markdown [hunk 2]
```diff
@@ -48,6 +48,7 @@ The following arguments are optional:
 * `name_prefix` - (Optional, Forces new resource) Creates a unique name beginning with the specified prefix. Conflicts with `name`.
 * `path` - (Optional, default "/") Path to the instance profile. For more information about paths, see [IAM Identifiers](https://docs.aws.amazon.com/IAM/latest/UserGuide/Using_Identifiers.html) in the IAM User Guide. Can be a string of characters consisting of either a forward slash (`/`) by itself or a string that must begin and end with forward slashes. Can include any ASCII character from the ! (\u0021) through the DEL character (\u007F), including most punctuation characters, digits, and upper and lowercase letters.
 * `role` - (Optional) Name of the role to add to the profile.
+* `tags` - Key-value mapping of tags for the IAM Instance Profile.
```
Comment (bflad) [inline website/docs/r/iam_instance_profile.html.markdown:51]:
```suggestion
* `tags` - (Optional) Map of resource tags for the IAM Instance Profile.
```

**Review by bflad — APPROVED** [review]
Looks great, thank you, @DrFaust92 🚀 

Output from acceptance testing in AWS Commercial:

```
--- PASS: TestAccAWSIAMInstanceProfile_withoutRole (24.71s)
--- PASS: TestAccAWSIAMInstanceProfile_disappears_role (25.87s)
--- PASS: TestAccAWSIAMInstanceProfile_disappears (26.09s)
--- PASS: TestAccAWSIAMInstanceProfile_namePrefix (28.30s)
--- PASS: TestAccAWSIAMInstanceProfile_basic (31.35s)
--- PASS: TestAccAWSIAMInstanceProfile_tags (44.67s)
```

Output from acceptance testing in AWS GovCloud (US):

```
--- PASS: TestAccAWSIAMInstanceProfile_withoutRole (32.70s)
--- PASS: TestAccAWSIAMInstanceProfile_disappears_role (34.17s)
--- PASS: TestAccAWSIAMInstanceProfile_disappears (34.40s)
--- PASS: TestAccAWSIAMInstanceProfile_basic (39.09s)
--- PASS: TestAccAWSIAMInstanceProfile_namePrefix (41.02s)
--- PASS: TestAccAWSIAMInstanceProfile_tags (69.02s)
```

**Pull request merged**